Type error message not localised enough

`test.ipso`
```
first : Array a -> (| Some : a, None : () |)
first =
  array.foldl
    (\result item ->
      case result of
        None () -> item
        Some _ -> result
    )
    (None ())

main : IO ()
main = io.pure ()
```

Expected output:

```
test.ipso:6:20: error: expected type "(| Some : a, None : () |)", got type "a"
  |
6 |         None () -> item
  |                    ^
```

Actual output:

```
test.ipso:3:3: error: expected type "Array a -> (| Some : a, None : () |)", got type "Array ((| None : (), Some : ?14 |)) -> (| None : (), Some : ?14 |)"
  |
3 |   array.foldl
  |   ^
```

What does GHC do?

```
first :: [a] -> Maybe a
first =
  foldl
    (\result item ->
      case result of
        Nothing -> item
        Just _ -> result
    )
    Nothing

---

1.hs:6:20: error:
    • Couldn't match expected type ‘Maybe a’ with actual type ‘a’
      ‘a’ is a rigid type variable bound by
        the type signature for:
          first :: forall a. [a] -> Maybe a
        at 1.hs:1:1-23
    • In the expression: item
      In a case alternative: Nothing -> item
      In the expression:
        case result of
          Nothing -> item
          Just _ -> result
    • Relevant bindings include
        item :: a (bound at 1.hs:4:14)
        result :: Maybe a (bound at 1.hs:4:7)
        first :: [a] -> Maybe a (bound at 1.hs:2:1)
  |
6 |         Nothing -> item
  |                    ^^^^
```

```
first :: [a] -> String
first arr =
  arr !! 0

---

2.hs:3:3: error:
    • Couldn't match type ‘a’ with ‘String’
      Expected: [String]
        Actual: [a]
      ‘a’ is a rigid type variable bound by
        the type signature for:
          first :: forall a. [a] -> String
        at 2.hs:1:1-22
    • In the first argument of ‘(!!)’, namely ‘arr’
      In the expression: arr !! 0
      In an equation for ‘first’: first arr = arr !! 0
    • Relevant bindings include
        arr :: [a] (bound at 2.hs:2:7)
        first :: [a] -> String (bound at 2.hs:2:1)
  |
3 |   arr !! 0
  |   ^^^
```

```
first :: [a] -> String
first =    
  (!! 0)

---

3.hs:3:4: error:
    • Couldn't match type ‘a’ with ‘String’
      Expected: [a] -> String
        Actual: [a] -> a
      ‘a’ is a rigid type variable bound by
        the type signature for:
          first :: forall a. [a] -> String
        at 3.hs:1:1-22
    • In the expression: !! 0
      In an equation for ‘first’: first = (!! 0)
    • Relevant bindings include
        first :: [a] -> String (bound at 3.hs:2:1)
  |
3 |   (!! 0)
  |    ^^^^
```

--- a/examples/25.dhall
+++ b/examples/25.dhall
@@ -5,10 +5,10 @@
   stdout = "",
   stderr =
     ''
-    25.ipso:15:9: error: expected type "String", got type "a"
+    25.ipso:15:24: error: expected type "Array String", got type "Array a"
        |
     15 |         (array.index 0 arr)
-       |         ^
+       |                        ^
     '',
   exitcode = 1
 }

--- a/examples/commentsErrors.ipso
+++ b/examples/commentsErrors.ipso
@@ -6,4 +6,3 @@ main : IO ()
 main = # comments shouldn't interfere with indentation
   print # really
     stdooout
-    "hello world!"

--- a/examples/comp_expr_error_2.dhall
+++ b/examples/comp_expr_error_2.dhall
@@ -5,7 +5,7 @@
   stdout = "",
   stderr =
     ''
-    comp_expr_error_2.ipso:4:15: error: expected type "IO ?5", got type "Int"
+    comp_expr_error_2.ipso:4:15: error: expected type "IO ?4", got type "Int"
       |
     4 |     bind x <- 666
       |               ^

--- a/examples/etaReductionTypeError.dhall
+++ b/examples/etaReductionTypeError.dhall
@@ -1,0 +1,14 @@
+{
+  description = "type error for an eta reduced function",
+  args = ["etaReductionTypeError.ipso"],
+  stdin = None Text,
+  stdout = "",
+  stderr =
+    ''
+    etaReductionTypeError.ipso:3:3: error: expected type "Array a -> String", got type "Array a -> a"
+      |
+    3 |   array.index 0
+      |   ^
+    '',
+  exitcode = 1
+}

--- a/examples/etaReductionTypeError.ipso
+++ b/examples/etaReductionTypeError.ipso
@@ -1,0 +1,6 @@
+first! : Array a -> String
+first! =
+  array.index 0
+
+main : IO ()
+main = io.pure ()

--- a/examples/issue377.dhall
+++ b/examples/issue377.dhall
@@ -1,0 +1,14 @@
+{
+  description = "https://github.com/LightAndLight/ipso/issues/377",
+  args = ["issue377.ipso"],
+  stdin = None Text,
+  stdout = "",
+  stderr =
+    ''
+    issue377.ipso:6:20: error: expected type "(| Some : a, None : () |)", got type "a"
+      |
+    6 |         None () -> item
+      |                    ^
+    '',
+  exitcode = 1
+}

--- a/examples/issue377.ipso
+++ b/examples/issue377.ipso
@@ -1,0 +1,12 @@
+first : Array a -> (| Some : a, None : () |)
+first =
+  array.foldl
+    (\result item ->
+      case result of
+        None () -> item
+        Some _ -> result
+    )
+    (None ())
+
+main : IO ()
+main = io.pure ()

--- a/ipso-typecheck/src/type_inference/mod.rs
+++ b/ipso-typecheck/src/type_inference/mod.rs
@@ -713,18 +713,36 @@ pub fn check(
                 },
             }
         }
-        syntax::Expr::App(fun, arg) => {
-            let in_ty = fresh_type_meta(&mut state.type_solutions, Kind::Type);
+        syntax::Expr::App(_fun, _arg) => {
+            fn unapply(
+                expr: &Spanned<syntax::Expr>,
+            ) -> (&Spanned<syntax::Expr>, Vec<&Spanned<syntax::Expr>>) {
+                match &expr.item {
+                    syntax::Expr::App(f, x) => {
+                        let (f, mut xs) = unapply(f.as_ref());
+                        xs.push(x.as_ref());
+                        (f, xs)
+                    }
+                    _ => (expr, Vec::new()),
+                }
+            }
+
+            let (fun, args) = unapply(expr);
+
+            let in_tys = args
+                .iter()
+                .map(|_| fresh_type_meta(&mut state.type_solutions, Kind::Type))
+                .collect::<Vec<_>>();
             let out_ty = fresh_type_meta(&mut state.type_solutions, Kind::Type);
 
             let fun = check(
                 env,
                 state,
                 fun,
-                &Type::mk_arrow(env.common_kinds, &in_ty, &out_ty),
+                &in_tys.iter().rev().fold(out_ty.clone(), |result, in_ty| {
+                    Type::arrow(env.common_kinds, in_ty.clone(), result)
+                }),
             )?;
-
-            let arg = check(env, state, arg, &in_ty)?;
 
             unification::unify(
                 env.as_unification_env(),
@@ -736,7 +754,13 @@ pub fn check(
             )
             .map_err(|error| Error::unification_error(env.source, position, error))?;
 
-            Ok(Expr::mk_app(fun, arg))
+            let args = args
+                .iter()
+                .zip(in_tys.iter())
+                .map(|(arg, in_ty)| check(env, state, arg, in_ty))
+                .collect::<Result<Vec<_>, _>>()?;
+
+            Ok(args.into_iter().fold(fun, Expr::mk_app))
         }
         syntax::Expr::Lam { args, body } => {
             check_duplicate_args(env.source, args)?;

